modules/internal/ChapelDynamicLoading.chpl

```diff
@@ -0,0 +1,797 @@
+        ref store = chpl_binaryInfoStore;
+        var shouldCreateNewEntry = true;
+
+        // Start by opening a system handle on LOCALE-0.
+        var err0;
+        var ptr0 = _systemDynLibOpen(path, err0);
+        defer { if ptr0 then _systemDynLibClose(ptr0); }
+
+        if err0 != nil || ptr0 == nil {
+          if err0 != nil then err = err0;
+          shouldCreateNewEntry = false;
+        }
+
+        // Check the LOCALE-0 store for an existing entry.
+        if shouldCreateNewEntry {
+          manage store.withReadLock() {
+            shouldCreateNewEntry = !store.get(ptr0, ret);
+          }
+        }
+
+        // None found, so initialize a new binary info.
+        if shouldCreateNewEntry {
+          var bin = new owned chpl_BinaryInfo(path, chpl_binaryInfoStore);
+
+          // Swap in the LOCALE-0 pointer and clear the local variable.
+          bin._systemPtrs[0] = ptr0;
+          ptr0 = nil;
+
+          var errBuf = new chpl_localBuffer(owned DynLibError?, numLocales);
+          var numErrs: atomic int = 0;
+
+          // Loop and attempt to open system handles on other locales.
+          if numLocales > 1 {
+            coforall loc in Locales[1..] with (ref errBuf) do on loc {
+              const n = (here.id: int);
+
+              var err;
+              var ptr = _systemDynLibOpen(path, err);
+              defer { if ptr then _systemDynLibClose(ptr); }
+
+              if err {
+                // The buffer is not wide! Assignment occurs on LOCALE-0.
+                on Locales[0] do errBuf[n] = err;
+                numErrs.fetchAdd(1);
+              } else if ptr {
+                // The buffer is not wide!
+                on Locales[0] do bin._systemPtrs[n] = ptr;
+                // Clear the pointer to avoid closing it.
+                ptr = nil;
+              } else {
+                // TODO: Construct an error instead.
+                halt('Failed to load system \'dlopen()\' pointer!');
+              }
+            }
+          }
+
+          if numErrs.read() {
+            // If there were errors, then report the first error.
+            // TODO: Consolidate the reported errors.
+            for i in 0..<errBuf.size {
+              if errBuf[i] {
+                err = errBuf[i];
+                break;
+              }
+            }
+          } else {
+            // Otherwise, try to add the newly created info to the store on
+            // LOCALE-0, or give up if another task already beat us to it.
+
+            for i in 0..<bin._systemPtrs.size {
+              // Make sure that all locales have succeeded in 'dlopen()'.
+              assert(bin._systemPtrs[i] != nil);
+            }
+
+            manage store.withWriteLock() {
+              var found = store.get(bin._systemPtrs[0], ret);
+              if !found {
+                ret = owned.release(bin);
+                var ok = store.set(ret!._systemPtrs[0], ret);
+                assert(ok);
+              }
```

Comment:
the asserts here make me slightly uneasy. Whats a scenario where the assert would be False, and could a user hit it? Make a `halt` guarded by `boundsChecks` will be more readable/give the user more context

Reply:
I don't think the asserts should be hit by construction, but they are there just in case. If I do end up triggering them I think it's OK to transform it into a halt at that time.